### DISCUSSION TITLE
Customize userinfo endpoint

### DISCUSSION THREAD

**Discussion by scriptbotprime** [keycloak/keycloak#27256] [Q&A]
I connected Apache Superset to Keycloak using OAuth. The SSO mechanism works well but it doesn't provide roles. Superset is a Flask App and I read in the [Flask Appbulder documentation](https://flask-appbuilder.readthedocs.io/en/latest/security.html) that it is possible to provide roles via the userinfo endpoint when they get provided in an atrribute called "role_keys". Is it somehow possible to customize the userinfo endpoint?

**Comment by jbman** [comment]
You can use a mapper at one of the client scope of your client defined at Keycloak.

1. Go to one of the scopes for your client "Superset" (e.g. dedicated scope):
![image](https://github.com/keycloak/keycloak/assets/789496/98006067-edc6-4dc7-9a0f-da407a6ad502)
2. Add mapper "client roles"
3. Configure the mapper: You can name the claim as you like, and add roles of a specific client this way.
![image](https://github.com/keycloak/keycloak/assets/789496/947c6984-d442-48c5-8968-2ebae45416d2)

Test it within the Keycloak admin console as described here: https://www.keycloak.org/docs/23.0.7/server_admin/#_client_scopes_evaluate

**Comment by scriptbotprime** [comment]
Thank you. I was able to set up the mapper. Tomorrow will try to connect superset.

**Comment by NemesisMate** [comment]
Turns out, you don't have just to add it to `userinfo` but set the `Token Claim Name` to `groups` since that's what Flask's AppBuilder looks for to map to `role_keys` Keycloak: https://github.com/dpgaspar/Flask-AppBuilder/blob/e2f7fdecffa770eba55083879c9eaf84bf1ecde3/flask_appbuilder/security/manager.py#L723
